extensions/amp-story/1.0/amp-story-tooltip.js

```diff
@@ -0,0 +1,401 @@
+import {setImportantStyles} from '../../../src/style';
+
+/**
+ * Minimum vertical space needed to position tooltip.
+ */
+const MIN_VERTICAL_SPACE = 56;
+
+/**
+ * Padding between tooltip and edges of screen.
+ */
+const EDGE_PADDING = 8;
+
+/**
+ * List of selectors that can trigger a tooltip.
+ * @const {!Array<string>}
+ */
+export const TOOLTIP_TRIGGERABLE_SELECTORS = ['a[href]'];
+
+/**
+ * @struct @typedef {{
+ *   tooltip: !Element,
+ *   buttonLeft: !Element,
+ *   buttonRight: !Element,
+ *   arrow: !Element,
+ * }}
+ */
+let tooltipElementsDef;
+
+/**
+ * Tooltip element triggered by clickable elements in the amp-story-grid-layer.
+ */
+export class AmpStoryTooltip {
+  /**
+   * @param {!Window} win
+   * @param {!Element} storyEl
+   */
+  constructor(win, storyEl) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Element} */
+    this.storyEl_ = storyEl;
+
+    /** @private {boolean} */
+    this.isBuilt_ = false;
+
+    /** @private {?Element} */
+    this.shadowRoot_ = null;
+
+    /** @private {?Element} */
+    this.tooltipOverlayEl_ = null;
+
+    /** @private {?Element} */
+    this.tooltip_ = null;
+
+    /** @private {?Element} */
+    this.tooltipArrow_ = null;
+
+    /** @private @const {!./amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = getStoreService(this.win_);
+
+    /** @private @const {!../../../src/service/resources-impl.Resources} */
+    this.resources_ = Services.resourcesForDoc(getAmpdoc(this.win_.document));
+
+    /**
+     * Clicked element producing the tooltip. Used to avoid building the same
+     * element twice.
+     * @private {?Element} */
+    this.clickedElem_ = null;
+
+    this.storeService_.subscribe(StateProperty.TOOLTIP_ELEMENT, elem => {
+      if (!elem) {
+        return;
+      }
+
+      if (!this.isBuilt_) {
+        this.storyEl_.appendChild(this.build_());
+      }
+
+      if (elem != this.clickedElem_) {
+        this.clickedElem_ = elem;
+        this.attachTooltipToEl_(elem);
+      }
+      this.onTooltipStateUpdate_(true);
+    });
+  }
+
+  /**
+   * Builds the tooltip overlay and appends it to the provided story.
+   * @private
+   */
+  build_() {
+    this.isBuilt_ = true;
+
+    this.shadowRoot_ = this.win_.document.createElement('div');
+
+    this.tooltipOverlayEl_ =
+      dev().assert(this.buildTemplate_(this.win_.document));
+    createShadowRootWithStyle(this.shadowRoot_, this.tooltipOverlayEl_, CSS);
+
+    this.tooltipOverlayEl_
+        .addEventListener('click', event => this.onOutsideClickableEls_(event));
+
+    this.storeService_.subscribe(StateProperty.UI_STATE, isDesktop => {
+      this.onUIStateUpdate_(isDesktop);
+    }, true /** callToInitialize */);
+
+    this.storeService_.subscribe(StateProperty.CURRENT_PAGE_ID, () => {
+      // Hide active tooltip when page switch is triggered by keyboard or
+      // desktop buttons.
+      this.closeTooltip_();
```

Comment:
Could you check if the tooltip is actually open before calling this method? Because it's executed on every navigation, our most memory/CPU intensive operation, it'd be safer to make sure it does nothing.